File: PE-diagram/src/nl/adaptivity/process/diagram/LayoutStepper.java
package nl.adaptivity.process.diagram;

import java.util.List;

import nl.adaptivity.diagram.Positioned;


@SuppressWarnings("unused")
public class LayoutStepper<T extends Positioned> {

  public void reportMove(DiagramNode<T> pNode, double newX, double newY) {
    // By default empty
  }

  public void reportPass(int pass) {
    // no implementation
  }

  public void reportLowest(List<? extends DiagramNode<T>> pNodes, DiagramNode<T> pNode) {
    // empty
  }

  public void reportHighest(List<? extends DiagramNode<T>> pNodes, DiagramNode<T> pNode) {
    // empty
  }

  public void reportRightmost(List<? extends DiagramNode<T>> pNodes, DiagramNode<T> pNode) {
    // empty
  }

  public void reportLeftmost(List<? extends DiagramNode<T>> pNodes, DiagramNode<T> pNode) {
    // empty
  }

  public void reportLayoutNode(DiagramNode<T> pNode) {
    // empty
  }

  public void reportMoveX(List<? extends DiagramNode<T>> pNodes, double offset) {
    for(DiagramNode<T> node: pNodes) {
      reportMove(node, node.getX()+offset, node.getY());
    }
  }

  public void reportMoveY(List<? extends DiagramNode<T>> pNodes, double offset) {
    for(DiagramNode<T> node: pNodes) {
      reportMove(node, node.getX(), node.getY()+offset);
    }
  }

  public void reportMinX(List<? extends DiagramNode<T>> pNodes, double offset) {
    // empty
  }

  public void reportMinY(List<? extends DiagramNode<T>> pNodes, double offset) {
    // empty
  }

  public void reportMaxX(List<? extends DiagramNode<T>> pNodes, double offset) {
    // empty
  }

  public void reportMaxY(List<? extends DiagramNode<T>> pNodes, double offset) {
    // empty
  }
  
  public void reportSiblings(DiagramNode<T> node, List<? extends DiagramNode<T>> nodes, boolean above) {
    // empty
  }

}

File: PE-diagram/src/main/kotlin/nl.adaptivity.process/diagram/LayoutStepper.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.diagram

import nl.adaptivity.diagram.Positioned

interface LayoutStepper<in T: Positioned> {
  fun reportMove(node: DiagramNode<T>, newX: Double, newY: Double) = Unit

  fun reportPass(pass: Int) = Unit

  fun reportLowest(nodes: List<DiagramNode<T>>, node: DiagramNode<T>) = Unit

  fun reportHighest(nodes: List<DiagramNode<T>>, node: DiagramNode<T>) = Unit

  fun reportRightmost(nodes: List<DiagramNode<T>>, node: DiagramNode<T>) = Unit

  fun reportLeftmost(nodes: List<DiagramNode<T>>, node: DiagramNode<T>) = Unit

  fun reportLayoutNode(node: DiagramNode<T>) = Unit

  fun reportMoveX(nodes: List<DiagramNode<T>>, offset: Double) =
    nodes.forEach { node -> reportMove(node, node.x + offset, node.y) }

  fun reportMoveY(nodes: List<DiagramNode<T>>, offset: Double) =
    nodes.forEach { node -> reportMove(node, node.x, node.y + offset) }

  fun reportMinX(nodes: List<DiagramNode<T>>, offset: Double) = Unit

  fun reportMinY(nodes: List<DiagramNode<T>>, offset: Double) = Unit

  fun reportMaxX(nodes: List<DiagramNode<T>>, offset: Double) = Unit

  fun reportMaxY(nodes: List<DiagramNode<T>>, offset: Double) = Unit

  fun reportSiblings(node: DiagramNode<T>, nodes: List<DiagramNode<T>>, above: Boolean) = Unit

}

open class AbstractLayoutStepper<T : Positioned> : LayoutStepper<T>
